src/simulation.jl

```diff
@@ -23,54 +27,127 @@ abstract type AbstractCall end
-    if isa(c.operator, AbstractArray)
-        operator = :(compose($operator))
+    #= if isa(operator, AbstractArray)
+        operator = Expr(:call, :∘, reverse(operator)...)
+    end =#
+    if(c.equality == :.=)
+        Expr(:call, :mul!, c.output, operator, c.input)
+    else
+        Expr(c.equality, c.output, Expr(:call, operator, c.input))
     end
-    return :($(c.output) = $operator($(c.input)))
 end
                 
 struct BinaryCall <: AbstractCall 
     operator
+    equality
     input1
     input2
     output
 end
 
+# TODO: After getting rid of AppCirc2, do we need this check?
 Base.Expr(c::BinaryCall) = begin
-    operator = c.operator
-    if isa(c.operator, AbstractArray)
+    #= if isa(c.operator, AbstractArray)
         operator = :(compose($(c.operator)))
-    end
-    return :($(c.output) = $operator($(c.input1), $(c.input2)))
+    end =#
+    return Expr(c.equality, c.output, Expr(:call, c.operator, c.input1, c.input2))
 end
 
 struct VarargsCall <: AbstractCall 
     operator
+    equality
     inputs
     output
 end
 
 Base.Expr(c::VarargsCall) = begin
-    operator = c.operator
-    if isa(c.operator, AbstractArray)
+    #= if isa(c.operator, AbstractArray)
         operator = :(compose($(c.operator)))
+    end =#
+    return Expr(c.equality, c.output, Expr(:call, c.operator, c.inputs...))
+end
+
+struct AllocVecCall <: AbstractCall 
+    name
+    form
+end
+
+# TODO: Check sizes on dual forms
+Base.Expr(c::AllocVecCall) = begin
+    resolved_form = @match c.form begin
+        :Form0 => :V
+        :Form1 => :E
+        :Form2 => :Tri
+        :DualForm0 => :Tri
+        :DualForm1 => :E
+        :DualForm2 => :V
+        _ => return :AllocVecCall_Error
     end
-    arglist = c.inputs
-    return :($(c.output) = $operator($(arglist...)))
+
+    :($(c.name) = Vector{Float64}(undef, nparts(mesh, $(QuoteNode(resolved_form)))))
 end
 
-function infer_states(d::AbstractNamedDecapode)
-    filter(parts(d, :Var)) do v
-        length(incident(d, v, :tgt)) == 0 &&
-        length(incident(d, v, :res)) == 0 &&
-        d[v, :type] != :Literal
+# TODO: Need to figure out how to deal with duals
+#= function get_form_number(d::SummationDecapode, var_id::Int)
+    type = d[var_id, :type]
+    if(type == :Form0)
+        return 0
+    elseif(type == :Form1)
+        return 1
+    elseif(type == :Form2)
+        return 2
+    end
+    return -1
+end
+
+# WARNING: This may not work if names are not unique, use above instead
+function get_form_number(d::SummationDecapode, var_name::Symbol)
+    var_id = first(incident(d, var_name, :name))
+    return get_form_number(d, var_id)
+end =#
+
+function is_form(d::SummationDecapode, var_id::Int)
+    type = d[var_id, :type]
+    if(type == :Form0 || type == :Form1 || type == :Form2 || 
+        type == :DualForm0 || type == :DualForm1 || type == :DualForm2)
+        return true
     end
+    return false
+end
+
+function is_form(d::SummationDecapode, var_name::Symbol)
```

Comment:
Maybe we could reduce some visual noise by converting some or all of these into one-liners?
e.g. is_literal(d::SummationDecapode, var_id::Int) = d[var_id, :type] == :Literal